### PR TITLE
Deprecation info support in RuntimeMetadataIR

### PR DESCRIPTION
**Pull request by pkhry**
### Description: 
* Adds `DeprecationStatusIR` enum to sp_metadata_ir.
Deprecation info for simple items.
* Adds `DeprecationInfoIR` enum to sp_metadata_ir.
It is a deprecation info for an enums/errors/calls. Contains `DeprecationStatusIR`.
Denotes full/partial deprecation of the type or its variants/calls
* Adds `deprecation_info` field to 
       - `RuntimeApiMetadataIR`
       - `RuntimeApiMethodMetadataIR`
       - `StorageEntryMetadataIR`
       - `PalletConstantMetadataIR`
       - `PalletCallMetadataIR`
       - `PalletMetadataIR`
       - `PalletEventMetadataIR`
       - `PalletErrorMetadataIR`

### Testing done: 
- Unit tests to check whether or not correct `note`/`since` texts are getting propagated to the metadata structs. 
- UI test to check for error message in case of incorrect attribute usage. 
There's also some test updates to make sure that deprecation attributes are getting propagated to the relevant structs.

see: #4098, Solution: A

### Examples of produced deprecation info metadata
They can be found in:
 - Tests for `frame-support`
 - hackmd link https://hackmd.io/@Zett98/Bys0YgbcR

### CODE REVIEW

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs [hunk 1]
```diff
@@ -58,7 +58,6 @@ pub fn expand_runtime_metadata(
 					#attr
 				}
 			});
-
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs:61]: nit: Unrelated change

File: substrate/frame/support/procedural/src/construct_runtime/parse.rs [hunk 2]
```diff
@@ -97,26 +97,29 @@ impl Parse for RuntimeDeclaration {
 		let pallets_token = pallets.token;
 
 		match convert_pallets(pallets.content.inner.into_iter().collect())? {
-			PalletsConversion::Implicit(pallets) =>
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/construct_runtime/parse.rs:100]: All changes of this file look unrelated. I think you'll need to configure the rustfmt to comply with substrate

File: substrate/primitives/api/proc-macro/src/utils.rs [hunk 3]
```diff
@@ -330,4 +391,32 @@ mod tests {
...
+		let simple_path: Attribute = parse_quote!(#[deprecated = #FIRST]);
+		let meta_list: Attribute = parse_quote!(#[deprecated(note = #FIRST)]);
+		let meta_list_with_since: Attribute =
+			parse_quote!(#[deprecated(note = #FIRST, since = #SECOND)]);
+		assert_eq!(
+			get_deprecation(&quote! { crate }, &[simple]).unwrap(),
+			quote! {crate_::metadata_id::DeprecationStatus::DeprecatedWithoutNote}
+		);
+		// assert_eq!(
+		// 	get_deprecation(&[simple_path]).unwrap(),
```
Comment (lexnv) [inline substrate/primitives/api/proc-macro/src/utils.rs:410]: dq: Should we uncomment thsese?

File: substrate/primitives/api/proc-macro/src/utils.rs [hunk 4]
```diff
@@ -330,4 +391,32 @@ mod tests {
...
+	#[test]
+	fn check_deprecated_attr() {
+		const FIRST: &'static str = "hello";
+		const SECOND: &'static str = "WORLD";
+
+		let simple: Attribute = parse_quote!(#[deprecated]);
+		let simple_path: Attribute = parse_quote!(#[deprecated = #FIRST]);
+		let meta_list: Attribute = parse_quote!(#[deprecated(note = #FIRST)]);
+		let meta_list_with_since: Attribute =
+			parse_quote!(#[deprecated(note = #FIRST, since = #SECOND)]);
```
Comment (lexnv) [inline substrate/primitives/api/proc-macro/src/utils.rs:404]: Would be interesting to see how this behaves with extra parameters

File: substrate/frame/support/test/tests/instance.rs [hunk 5]
```diff
@@ -441,6 +441,7 @@ fn expected_metadata() -> PalletStorageMetadataIR {
 				ty: StorageEntryTypeIR::Plain(scale_info::meta_type::<u32>()),
 				default: vec![0, 0, 0, 0],
 				docs: vec![],
+				deprecation_info: sp_metadata_ir::DeprecationStatus::NotDeprecated,
```
Comment (lexnv) [inline substrate/frame/support/test/tests/instance.rs:444]: I would add a few tests similar to these (with/without note=.. since=..):
- check deprecation info for storage entries
- check deprecation info for runtime APIs for individual methods and for the trait itself
- check deprecation info for constants 

This way we'll also have a good example and see the API in action from the user's perspective

**Comment by lexnv** [comment]
Nice work here @pkhry! Certainly off to a good start! 👍 

Left a few thoughts after the initial review, mainly:
- would be good to keep the changes related to this PR only (ie adjust the rustfmt)
- a few more tests to illustrate how this feels in practice for the user experience (also to validate the approach for all deprecation included calls / storage etc)
- would be interesting also to include a negative test that fails to compile because `pallet::deprecated` includes unsupported fields (ie fields that are not empty; "note" or "since")
- make the CI happy (to double check the tests for now, everything should pass since changes are made in isolation)


From the CI step: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/6513912:
```
error: unused import: `proc_macro::TokenStream`
  --> substrate/frame/support/procedural/src/pallet/parse/config.rs:20:5
   |
20 | use proc_macro::TokenStream;
   |     ^^^^^^^^^^^^^^^^^^^^^^^
```

After that we can have a look at exposing deprecation info for events / errors (maybe in a separate PR), your initial approach of using a `BTreeMap` sounds good offhand, but I would consolidate this first. Let me know if you have any questions here 🙏

File: substrate/frame/support/src/tests/mod.rs [hunk 6]
```diff
@@ -101,18 +101,19 @@ pub mod frame_system {
...
-				return Err(Error::<T>::FailedTask.into())
+				return Err(Error::<T>::FailedTask.into());
 			}
 
 			Ok(().into())
 		}
 	}
 
 	#[pallet::storage]
+	#[deprecated]
```
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/frame/support/src/tests/mod.rs:116]: hmm, but that would introduce the need to use two separate deprecation attributes, one for metadata and one for rustc deprecation warnings. Or that's totally ok?

File: substrate/frame/support/test/tests/instance.rs (hunk 5 above)
Comment (pkhry) [reply substrate/frame/support/test/tests/instance.rs:444]: added more testcases for `StorageEntryMetadataIr` and constants metadata

File: substrate/primitives/api/proc-macro/src/utils.rs (hunk 4 above)
Comment (pkhry) [reply substrate/primitives/api/proc-macro/src/utils.rs:403]: it will just ignore them

File: substrate/primitives/api/proc-macro/src/utils.rs (hunk 3 above)
Comment (pkhry) [reply substrate/primitives/api/proc-macro/src/utils.rs]: done

File: substrate/frame/support/procedural/src/construct_runtime/parse.rs (hunk 2 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/construct_runtime/parse.rs]: didn't notice that it was relying on nightly settings, corrected

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs (hunk 1 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs]: removed it

File: substrate/primitives/api/proc-macro/src/utils.rs (hunk 4 above)
Comment (lexnv) [reply substrate/primitives/api/proc-macro/src/utils.rs:402]: Do you think we can turn that into a compile error? And specify inside the message the forms we support?

File: substrate/frame/support/src/tests/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lexnv) [reply substrate/frame/support/src/tests/mod.rs:122]: I think that would make it a bit clear that the `deprecation` tag is collected into the metadata. Then, the `pallet::deprecated` could insert under the hood the `deprecated` attribute to the type, and propagate `DeprecationInfo` to the metadata. Up to you in the end, I have no strong pref here :D

File: substrate/primitives/api/proc-macro/src/utils.rs (hunk 4 above)
Comment (pkhry) [reply substrate/primitives/api/proc-macro/src/utils.rs:402]: it's a compiler error already, that's one of the reason why i decided to use rust's deprecated. But the parsing now doesnt fail if attribute is malformed(not sure how useful it is tho)

File: substrate/frame/support/src/tests/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/frame/support/src/tests/mod.rs:122]: what about `sp-api-proc-macro` stuff? should it also be `pallet::deprecated` there or for example `api::deprecated`?

[the remark this answers is not in the document]
Comment (lexnv) [reply substrate/frame/support/src/tests/mod.rs:122]: I guess we can go with the default `deprecation` tag for now, I would focus on the metadata collection first, we can come back and change it later if needed

File: substrate/frame/support/procedural/src/pallet/expand/constants.rs [hunk 7]
```diff
@@ -45,7 +47,9 @@ pub fn expand_constants(def: &mut Def) -> proc_macro2::TokenStream {
 	let config_consts = def.config.consts_metadata.iter().map(|const_| {
 		let ident = &const_.ident;
 		let const_type = &const_.type_;
-
+		let deprecation_info =
+			crate::deprecation::get_deprecation(&quote::quote! { #frame_support }, &const_.attrs)
+				.expect("invalid deprecation attribute");
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/constants.rs:52]: Is this checked in the parsing step or somewhere?
macros should not panic on wrong input

File: substrate/frame/support/procedural/src/pallet/expand/constants.rs [hunk 8]
```diff
@@ -56,11 +60,15 @@ pub fn expand_constants(def: &mut Def) -> proc_macro2::TokenStream {
...
 			metadata_name: None,
+			deprecation_info: quote::quote! { #deprecation_info },
 		}
 	});
 
 	let extra_consts = def.extra_constants.iter().flat_map(|d| &d.extra_constants).map(|const_| {
 		let ident = &const_.ident;
+		let deprecation_info =
+			crate::deprecation::get_deprecation(&quote::quote! { #frame_support }, &const_.attrs)
+				.expect("invalid deprecation attribute");
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/constants.rs:71]: same as above

File: substrate/frame/support/procedural/src/pallet/expand/constants.rs [hunk 9]
```diff
@@ -56,11 +60,15 @@ pub fn expand_constants(def: &mut Def) -> proc_macro2::TokenStream {
 				#frame_support::__private::codec::Encode::encode(&value)
 			),
 			metadata_name: None,
+			deprecation_info: quote::quote! { #deprecation_info },
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/constants.rs:63]:
```suggestion
			deprecation_info,
```

File: substrate/frame/support/procedural/src/pallet/expand/constants.rs [hunk 10]
```diff
@@ -71,6 +79,7 @@ pub fn expand_constants(def: &mut Def) -> proc_macro2::TokenStream {
 				#frame_support::__private::codec::Encode::encode(&value)
 			),
 			metadata_name: const_.metadata_name.clone(),
+			deprecation_info: quote::quote! { #deprecation_info },
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/constants.rs:82]:
```suggestion
			deprecation_info,
```

File: substrate/frame/support/src/tests/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (thiolliere) [reply substrate/frame/support/src/tests/mod.rs:122]: Using the native rust `deprecated` is a bit similar to that we are using the native rust doc attribute.

Also the information of `deprecated` could be added to scale-info as well in the future.

**Review by thiolliere — COMMENTED** [review]
It could be great to have at least 1 compilation error test. To ensure that if user writes something like `#[deprecated(note = 3)]` then there is a good error message.

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+	}
+}
+
+/// collects deprecation attribute if its present.
+pub fn get_deprecation(path: &TokenStream, attrs: &[syn::Attribute]) -> Result<TokenStream> {
+	attrs
+		.iter()
+		.find(|a| a.path().is_ident("deprecated"))
+		.map(|a| parse_deprecated_meta(path, a))
+		.unwrap_or_else(|| {
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/deprecation.rs:79]: How could we make the difference here between a missing deprecation atttribute and the error that was produced during parsing the attribute?
- Errors here should translate above into a `compile_err` with sufficient information for the user to rectify it. Maybe turn the `parse_deprecated_meta` into a `Result<Option<TokenStream>>`? 
- For errors propagate them directly and if the get a `Ok(None)` we can trun that into a `Ok(NotDeprecated)` here

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+		}) => {
+			// #[deprecated = "lit"]
+			let doc = quote! { #path::__private::metadata_ir::DeprecationStatus::Deprecated { note: #lit, since: None } };
+			Ok(doc)
+		},
+		Meta::Path(_) => {
+			// #[deprecated]
+			Ok(quote! { #path::__private::metadata_ir::DeprecationStatus::DeprecatedWithoutNote })
+		},
+		_ => Err(Error::new(attr.span(), "Invalid deprecation attribute")),
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/deprecation.rs:69]: Nit: Maybe we can provide a bit more info the user since these will be turned into compile_err? Maybe "Accepted formats are: #[deprecated], #[deprecated( note since)], #[deprecated = "info" ]"

Comment (pkhry) [reply substrate/frame/support/procedural/src/deprecation.rs:69]: user should get the proper explanation in those cases from the rust compiler itself. added a ui test as an example of this output

File: substrate/frame/support/procedural/src/pallet/expand/constants.rs (hunk 9 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/constants.rs]: done

File: substrate/frame/support/procedural/src/pallet/expand/constants.rs (hunk 10 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/constants.rs]: done

File: substrate/primitives/api/proc-macro/src/runtime_metadata.rs [hunk 13]
```diff
@@ -131,20 +131,23 @@ pub fn generate_decl_runtime_metadata(decl: &ItemTrait) -> TokenStream2 {
...
 				docs: #docs,
+				deprecation_info: #deprecation,
 			}
 		));
 	}
 
 	let trait_name_ident = &decl.ident;
 	let trait_name = trait_name_ident.to_string();
 	let docs = collect_docs(&decl.attrs, &crate_);
+	let deprecation = crate::utils::get_deprecation(&crate_, &decl.attrs).unwrap();
```
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/api/proc-macro/src/runtime_metadata.rs]: done

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+fn parse_deprecated_meta(path: &TokenStream, attr: &syn::Attribute) -> Result<TokenStream> {
+	match &attr.meta {
+		Meta::List(meta_list) => {
+			let parsed = meta_list
+				.parse_args_with(Punctuated::<MetaNameValue, Token![,]>::parse_terminated)
+				.map_err(|e| Error::new(attr.span(), e.to_string()))?;
+			let (note, since) = parsed.iter().try_fold((None, None), |mut acc, item| {
+				let value = match &item.value {
+					Expr::Lit(ExprLit { lit: lit @ Lit::Str(_), .. }) => Ok(lit),
+					_ => Err(Error::new(attr.span(), "Invalid deprecation attribute")),
```
Comment (gui1117) [inline substrate/frame/support/procedural/src/deprecation.rs:34]: maybe better?

```suggestion
					_ => Err(Error::new(item.value.span(), "Invalid deprecation attribute: expect string literal")),
```

File: substrate/frame/support/procedural/src/deprecation.rs (hunk 12 above)
Comment (gui1117) [reply substrate/frame/support/procedural/src/deprecation.rs:69]: But the first error would still be the one from the macro I think.

Best would be to be more descriptive:
```diff
@@ -66,7 +66,8 @@ fn parse_deprecated_meta(path: &TokenStream, attr: &syn::Attribute) -> Result<To
                        // #[deprecated]
                        Ok(quote! { #path::__private::metadata_ir::DeprecationStatus::DeprecatedWithoutNote })
                },
-               _ => Err(Error::new(attr.span(), "Invalid deprecation attribute")),
+               Meta::NameValue(name_value) =>
+                       Err(Error::new(name_value.value.span(), "Invalid deprecation attribute: expect string literal")),
        }
 }
 
```

Or just give the helper message from rust:

```
help: the following are the possible correct uses
    |
410 |     #[deprecated = "reason"]
    |
410 |     #[deprecated(/*opt*/ since = "version", /*opt*/ note = "reason")]
    |
410 |     #[deprecated]
    |
```

File: substrate/primitives/api/test/tests/ui/deprecation_info.stderr [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+   |
+22 |         #[deprecated(since = 5)]
+   |         ^
+
+error: Invalid deprecation attribute
+  --> tests/ui/deprecation_info.rs:24:3
+   |
+24 |         #[deprecated = 5]
+   |         ^
+
```
Comment (gui1117) [inline substrate/primitives/api/test/tests/ui/deprecation_info.stderr:18]: as suggested those could give more information as they are the first to be show up to user.

**Review by gui1117 — COMMENTED** [review]
Some suggestions for the error message. otherwise code is good

File: substrate/primitives/api/test/tests/ui/deprecation_info.stderr (hunk 15 above)
Comment (pkhry) [reply substrate/primitives/api/test/tests/ui/deprecation_info.stderr]: pushed the suggested change

File: substrate/frame/support/procedural/src/deprecation.rs (hunk 14 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/deprecation.rs]: pushed the suggested change

File: substrate/frame/support/procedural/src/deprecation.rs (hunk 12 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/deprecation.rs]: pushed the suggested change

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+	match &attr.meta {
+		Meta::List(meta_list) => {
+			let parsed = meta_list
+				.parse_args_with(Punctuated::<MetaNameValue, Token![,]>::parse_terminated)
+				.map_err(|e| Error::new(attr.span(), e.to_string()))?;
+			let (note, since) = parsed.iter().try_fold((None, None), |mut acc, item| {
+				let value = match &item.value {
+					Expr::Lit(ExprLit { lit: lit @ Lit::Str(_), .. }) => Ok(lit),
+					_ => Err(Error::new(
+						attr.span(),
```
Comment (gui1117) [inline substrate/frame/support/procedural/src/deprecation.rs:49]: you could also use `item.value.span()` if you want

**Review by gui1117 — APPROVED** [review]
Looks good to me

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{
+	punctuated::Punctuated, spanned::Spanned, Error, Expr, ExprLit, Lit, Meta, MetaNameValue,
+	Result, Token, Variant,
+};
+fn deprecation_msg_formatter(msg: &str) -> String {
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/deprecation.rs:24]: Nice! I like this format!

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+				if item.path.is_ident("note") {
+					acc.0.replace(value);
+				} else if item.path.is_ident("since") {
+					acc.1.replace(value);
+				} else {
+				};
+				Ok::<(Option<&syn::Lit>, Option<&syn::Lit>), Error>(acc)
+			})?;
+			note.map_or_else(
+				|| Err(Error::new(attr.span(), 						deprecation_msg_formatter(
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/deprecation.rs:64]: nit: Do we need to run cargo fmt here? Seems a bit odd with that many spaces?

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+		})
+}
+
+/// collects deprecation attribute if its present for enum-like types
+pub fn get_deprecation_enum<'a>(
+	path: &TokenStream,
+	parent_attrs: &[syn::Attribute],
+	children_attrs: impl Iterator<Item = (u8, &'a [syn::Attribute])>,
+) -> Result<TokenStream> {
+	fn parse_deprecation(
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/deprecation.rs:115]: nit: This seems familiar with `fn get_deprecation`. Do you think we can place the `fn parse_deprecation` outside this function, then make the `get_deprecation` call into that?

```
fn get_deprecation
  parse_deprecation.map(|opt| opt.unwrap_or_else( NotDeprecated) )
```

**Review by lexnv — APPROVED** [review]
Looks good to me! Nice work here!

Overall the code logic looks good! I do like the new fields introduced in the `metadata-ir/types.rs`, after some time we could introduce them to the `frame-metadata` as well (after this PR reviewed and merged). Then we can expose the metadata v16 as unstable for the end users. This PR already does the heavy lifting in exposing the deprecation info! 🙏 

I think there are some smaller bits and pieces, like cargo clippy.

**Comment by kianenigma** [comment]
I feel like I wrote this in the last few days somewhere, but my memoryo is failing me 🙈  better safe than sorry, so here it goes again: 

The main user of this is PJS and PAPI right? as in, they can now show the deprecated stuff to the TS users. Let's communicate to them please, esp. PAPI because they are external.

File: substrate/frame/support/procedural/src/deprecation.rs (hunk 19 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/deprecation.rs]: good catch, changed as per suggestion

File: substrate/frame/support/procedural/src/deprecation.rs (hunk 18 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/deprecation.rs]: seems to be related to cargo fmt itself

Comment (gui1117) [reply substrate/frame/support/procedural/src/deprecation.rs]: it feels like fmt doesn't format proc-macro crates? I think if you fix manually, then cargo fmt won't change it.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/metadata-v16-updates/9557/1

**Comment by pkhry** [comment]
bot fmt

**Review by kianenigma — APPROVED** [review]
Approving on behalf of @gui1117

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{
+	punctuated::Punctuated, spanned::Spanned, Error, Expr, ExprLit, Lit, Meta, MetaNameValue,
+	Result, Token, Variant,
+};
```
Comment (bkchr) [inline substrate/frame/support/procedural/src/deprecation.rs:23]:
```suggestion
};

```

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+		"{msg}\n{helper}",
+		msg = msg,
+		helper = r#"help: the following are the possible correct uses
+|
+|     #[deprecated = "reason"]
+|
+|     #[deprecated(/*opt*/ since = "version", /*opt*/ note = "reason")]
+|
+|     #[deprecated]
+|"#
```
Comment (bkchr) [inline substrate/frame/support/procedural/src/deprecation.rs:35]:
```suggestion
	format!(
		r#"{msg}
		help: the following are the possible correct uses
|
|     #[deprecated = "reason"]
|
|     #[deprecated(/*opt*/ since = "version", /*opt*/ note = "reason")]
|
|     #[deprecated]
|"#
```

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+						if item.path.is_ident("index") {
+							acc.replace(*val);
+						}
+					})
+				};
+				acc
+			})
+		})
+		.unwrap_or(index);
+	index
```
Comment (bkchr) [inline substrate/frame/support/procedural/src/deprecation.rs:192]:
```suggestion
	item
		.attrs
		.iter()
		.find(|attr| attr.path().is_ident("codec"))
		.and_then(|attr| {
			if let Meta::List(meta_list) = &attr.meta {
				meta_list
					.parse_args_with(Punctuated::<MetaNameValue, syn::Token![,]>::parse_terminated)
					.ok()
			} else {
				None
			}
		})
		.and_then(|parsed| {
			parsed.iter().fold(None, |mut acc, item| {
				if let Expr::Lit(ExprLit { lit: Lit::Int(num_lit), .. }) = &item.value {
					num_lit.base10_parse::<u8>().iter().for_each(|val| {
						if item.path.is_ident("index") {
							acc.replace(*val);
						}
					})
				};
				acc
			})
		})
		.unwrap_or(index)
```

File: substrate/frame/support/procedural/src/pallet/expand/storage.rs [hunk 23]
```diff
@@ -418,11 +417,14 @@ pub fn expand_storages(def: &mut Def) -> proc_macro2::TokenStream {
...
 		let cfg_attrs = &storage.cfg_attrs;
-
+		let deprecation =
+			crate::deprecation::get_deprecation(&quote::quote! { #frame_support }, &storage.attrs)
+				.unwrap_or_else(syn::Error::into_compile_error);
 		quote::quote_spanned!(storage.attr_span =>
 			#(#cfg_attrs)*
 			{
 				<#full_ident as #frame_support::storage::StorageEntryMetadataBuilder>::build_metadata(
+					#deprecation,
```
Comment (bkchr) [inline substrate/frame/support/procedural/src/pallet/expand/storage.rs:427]: This will also lead to other compile errors.

File: substrate/primitives/metadata-ir/src/types.rs [hunk 24]
```diff
@@ -451,3 +463,63 @@ impl IntoPortable for OuterEnumsIR {
 		}
 	}
 }
+
+/// Deprecation status for an entry inside MetadataIR
+#[derive(Clone, PartialEq, Eq, Encode, Debug)]
+pub enum DeprecationStatusIR<T: Form = MetaForm> {
+	/// Entry is not deprecated
+	NotDeprecated,
```
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs:471]: I've decided to make this state a part of the variant instead of putting other two inside an option. I can move it to option if needed

File: substrate/primitives/metadata-ir/src/types.rs [hunk 25]
```diff
@@ -451,3 +463,63 @@ impl IntoPortable for OuterEnumsIR {
...
+#[derive(Clone, PartialEq, Eq, Encode, Debug)]
+pub enum DeprecationInfoIR<T: Form = MetaForm> {
+	/// Type is not deprecated
+	NotDeprecated,
+	/// Entry is fully deprecated.
+	FullyDeprecated(DeprecationStatusIR<T>),
+	/// Entry is partially deprecated.
+	/// For Errors and Events this means that only some of the variants are deprecated
+	/// For Calls only certain call indexes are deprecated
+	PartiallyDeprecated(BTreeMap<Compact<u8>, DeprecationStatusIR<T>>),
```
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs:508]: It will be a compile error as having both is nonsensical. Initially we would just ignore internal deprecation attributes if enum itself had it

File: substrate/primitives/metadata-ir/src/types.rs [hunk 26]
```diff
@@ -451,3 +463,63 @@ impl IntoPortable for OuterEnumsIR {
...
+	}
+}
+/// Deprecation info for an enums/errors/calls.
+/// Denotes full/partial deprecation of the type
+#[derive(Clone, PartialEq, Eq, Encode, Debug)]
+pub enum DeprecationInfoIR<T: Form = MetaForm> {
+	/// Type is not deprecated
+	NotDeprecated,
+	/// Entry is fully deprecated.
+	FullyDeprecated(DeprecationStatusIR<T>),
```
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs:504]: Not sure variants would be a good fit as we also cover `Call` with it

File: substrate/primitives/metadata-ir/src/types.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply substrate/primitives/metadata-ir/src/types.rs:508]: If Rust supports this, it is not "nonsensical"

File: substrate/primitives/metadata-ir/src/types.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply substrate/primitives/metadata-ir/src/types.rs:504]: `Call` is just an enum. Or what do you mean?

[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs:504]: i'm talking about `PalletCallMetadataIR` produced by impls annotated by `#[pallet::call]`. Its also using indexes, but its not an enum

**Comment by pkhry** [comment]
bot fmt

File: substrate/primitives/api/proc-macro/src/runtime_metadata.rs [hunk 27]
```diff
@@ -131,20 +131,25 @@ pub fn generate_decl_runtime_metadata(decl: &ItemTrait) -> TokenStream2 {
...
+		let deprecation = crate::utils::get_deprecation(&crate_, &method.attrs)
+			.unwrap_or_else(syn::Error::into_compile_error);
 		methods.push(quote!(
 			#( #attrs )*
 			#crate_::metadata_ir::RuntimeApiMethodMetadataIR {
 				name: #method_name,
 				inputs: #crate_::vec![ #( #inputs, )* ],
 				output: #output,
 				docs: #docs,
+				deprecation_info: #deprecation,
```
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/api/proc-macro/src/runtime_metadata.rs:143]: I'm having trouble understanding this, could you please elaborate further?

File: substrate/frame/support/procedural/src/deprecation.rs [hunk 28]
```diff
@@ -0,0 +1,193 @@
...
+				if item.path.is_ident("note") {
+					acc.0.replace(value);
+				} else if item.path.is_ident("since") {
+					acc.1.replace(value);
+				} else {
+				};
+				Ok::<(Option<&syn::Lit>, Option<&syn::Lit>), Error>(acc)
+			})?;
+			note.map_or_else(
+		  || Err(Error::new(attr.span(), 						deprecation_msg_formatter(
```
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/frame/support/procedural/src/deprecation.rs]: should be fixed now

File: substrate/primitives/metadata-ir/src/types.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply substrate/primitives/metadata-ir/src/types.rs:504]: Internally this maps to an enum. This is the reason you have there these indices.

**Comment by pkhry** [comment]
bot fmt

File: substrate/primitives/metadata-ir/src/types.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs]: renamed as suggested

File: substrate/primitives/api/proc-macro/src/runtime_metadata.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/api/proc-macro/src/runtime_metadata.rs:145]: Now code does an early return if deprecation_attribute error is encountered

File: substrate/primitives/api/proc-macro/src/runtime_metadata.rs [hunk 29]
```diff
@@ -174,6 +179,7 @@ pub fn generate_decl_runtime_metadata(decl: &ItemTrait) -> TokenStream2 {
 					name: #trait_name,
 					methods: #crate_::vec![ #( #methods, )* ],
 					docs: #docs,
+					deprecation_info: #deprecation,
```
[the remark this answers is not in the document]
Comment (pkhry) [reply substrate/primitives/api/proc-macro/src/runtime_metadata.rs:186]: Now code does an early return if deprecation_attribute error is encountered

File: substrate/frame/support/procedural/src/pallet/expand/storage.rs (hunk 23 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/storage.rs:432]: Now code does an early return if deprecation_attribute error is encountered

**Comment by pkhry** [comment]
bot update-ui

**Review by bkchr — APPROVED** [review]
